### ISSUE TITLE
cacti/graph_view.php -> Preview Mode -> problem with search_text_field (regex)

### ISSUE THREAD

**Issue by dabeani** [Cacti/cacti#1161]
Hi,

If you navigate to the graph_view.php - and klick on the right side the "preview view" button, and enter in the search field some query with regex (something which conatins '[' and ']', like '[0-9]'), you will see the correct result. But if you wait for the auto-refresh you will get "no graph" shown. What i figured out so far, the sourcecode of the graph_view.php (compiled source after entering the search string) will show a variable (var refreshPage=...) where you can see the missing '[',']'... So there is some problem with the transformation of the brackets.

if you need more information, let me know..

**Comment by netniV** [comment]
I'll take a look but this may be down to the field being "sanitized" so that it doesn't contain script code.

**Comment by dabeani** [comment]
thank you!
will be nice to use the brackets because some type of regex needs this.

**Comment by netniV** [comment]
The problem is, allowing the passing of regex brackets rather than just basic regex can result in issues where the regex engine can be overtaxed by a malicious parameter.  I'll see if I can see anything obvious, but the team may not want that functionality changed for that reason. 

One thing, I don't actually see an auto refresh on that page, do you have that setup to use your browser or an addon auto-refresh?

**Comment by dabeani** [comment]
mhm, okay - bad. Will be nice to have a full-usable regex search field. Hope there will be some special thing to allow that :).
You will find the Page-Refresh in the User-Settings.
Console / Users / User Settings -> Page Refresh (5 Minutes)

**Comment by netniV** [comment]
Yeah, that's the refresh I thought you were referring to, just wanted to be sure.  From what I can see, the page refresh (or timeout) reload the entire page.  It uses the $_SERVER['REQUEST_URI'] as part of that function so will have to check whether the AJAX call is receiving that correctly.

**Comment by netniV** [comment]
It's as I suspected, the field is being stripped for any potential code values.  I set my test user to 15 second refresh time, then looked at the URL when I manually clicked search plus that on the refresh:

```
[02/Jan/2018:15:08:11 +0000] "GET /graph_view.php?action=preview&rfilter=db[1,2]&host_id=-1&columns=3&graphs=100&graph_template_id=0&thumbnails=true&header=false HTTP/1.1" 200 "/graph_view.php?action=preview" "

[02/Jan/2018:15:08:04 +0000] "GET /graph_view.php?action=preview&rfilter=db1,2&host_id=-1&columns=3&graphs=100&graph_template_id=0&thumbnails=true&header=false&header=false&header=false&header=false&header=false&header=false&header=false&header=false HTTP/1.1" 200 6123 "/graph_view.php?action=preview"
```

The other curious thing is why header has been added so many times to the URI, but I suspect that this is because it wasn't actually checked for before setting it again.

**Comment by dabeani** [comment]
jep, sounds like the same as i already saw in the sourcecode :).
it looks like we can not change things to use the brackets - so i think i should change that brackets to some Solution without brackets...

if someone have a better idea or knows how to solve that issue, that will be great!
thx so far for the help and debugging.

**Comment by netniV** [comment]
Well, not strictly true.  If it is allowing it on the AJAX call, it should also handle it on the refresh.  I think the problem is that to prevent XSS, the refresh option automatically parses them out.  I'll see if I can create a "quick fix" for you, but it may be one that doesn't make it into the package as a whole so you may need to re-apply after any upgrade.

**Comment by netniV** [comment]
Actually, I've looked into this further and the RFILTER parameter is defined as requiring REGEX validation suggesting that it should be passed around properly.  I will definitely dig further into this now :)

> include/html_graph.php, Line 108
```

                'rfilter' => array(
                        'filter' => FILTER_VALIDATE_IS_REGEX,
                        'pageset' => true,
                        'default' => '',
                        ),
```

**Comment by netniV** [comment]
Further debugging shows that there could be an issue as the rfilter is being used with get_request_var rather than get_filter_request_var.  However, even after attempting that, the location header is the part that is actually being changed and I'm not sure but what yet! :)

**Comment by cigamit** [comment]
The solution is to encode the regex into a base64 encoded string, and then properly parse it on the backend before calling the regex validation function.  Solution should go in later this evening.  This should be done where ever there is a regex possible and a page refresh is involved.

**Comment by netniV** [comment]
The problem appears to be with the applySkin() function as it uses the requestURI variable.  I know for a fact that this was discussed recently as being sanitized so it will remove any extraneous characters such as those found in the regex of a filter :)

Issue #1137 has some basic discussions around that field being sanitized. @cigamit or @ronytomen  may have a comment on what could be done about that.

I confirmed this by adding the following code to include/layout.js:

```
function pushState(myTitle, myHref) {
        // show the stacktrace of where the state was changed
        // If refresh page is low, this will be shown often
        console.log(myHref + '\n' + (new Error().stack));
```

This recorded that as the page was loaded, the pageState was updated to reflect the following URL despite my original URL having db[1,2]

```
/graph_view.php?action=preview&rfilter=db1,2&host_id=-1&columns=2&graphs=10&graph_template_id=0&thumbnails=true&header=false
Error
    at pushState (https://server//include/layout.js:1720:31)
    at applySkin (https://server//include/layout.js:551:3)
    at Object.success (https://server//include/layout.js:1704:3)
    at fire (https://server//include/js/jquery.js:3305:31)
    at Object.fireWith [as resolveWith] (https://server//include/js/jquery.js:3435:7)
    at done (https://server//include/js/jquery.js:9242:14)
    at XMLHttpRequest.<anonymous> (https://server//include/js/jquery.js:9484:9)
layout.js:1720 https://server//graph_view.php
Error
    at pushState (https://server//include/layout.js:1720:31)
    at Object.success (https://server//include/layout.js:1706:3)
    at fire (https://server//include/js/jquery.js:3305:31)
    at Object.fireWith [as resolveWith] (https://server//include/js/jquery.js:3435:7)
    at done (https://server//include/js/jquery.js:9242:14)
    at XMLHttpRequest.<anonymous> (https://server//include/js/jquery.js:9484:9)
```

**Comment by netniV** [comment]
I see @cigamit commented just before my last update as I was typing it out :)

**Comment by cigamit** [comment]
Please test the files in the latest commit.  Remember to reload the layout.js from cache using F5 on Firefox and IE.  For Chrome, it's best to open the file directly, and then using F5 with the JavaScript file open.

**Comment by netniV** [comment]
That appears to be working for my local testing.  The only thing I did notice was an uncaught exception occurring

```
layout.js:2028 Uncaught ReferenceError: applyFilter is not defined
    at HTMLFormElement.<anonymous> (layout.js:2028)
    at HTMLFormElement.dispatch (jquery.js:5201)
    at HTMLFormElement.elemData.handle (jquery.js:5009)
(anonymous) @ layout.js:2028
dispatch @ jquery.js:5201
elemData.handle @ jquery.js:5009
```

This appears to be with this event code:

```
        $('#form_graph_view').unbind('submit').on('submit', function(event) {
                event.preventDefault();
                event.stopPropagation();
                applyFilter();
        });
```

_This may be one line out as my layout.js still had my stacktrace recording in it so I could check through the logs_

```
/graph_view.php?newtheme=1
Error
    at pushState (https://server//include/layout.js:1724:38)
    at applySkin (https://server//include/layout.js:556:3)
    at HTMLDocument.<anonymous> (https://server//include/layout.js:1568:2)
    at mightThrow (https://server//include/js/jquery.js:3570:29)
    at process (https://server//include/js/jquery.js:3638:12)

Uncaught ReferenceError: applyFilter is not defined
    at HTMLFormElement.<anonymous> (layout.js:2028)
    at HTMLFormElement.dispatch (jquery.js:5201)
    at HTMLFormElement.elemData.handle (jquery.js:5009)
/graph_view.php?action=preview&rfilter=W2Zddw==&host_id=-1&columns=2&graphs=10&graph_template_id=0&thumbnails=true&header=false
Error
    at pushState (https://server//include/layout.js:1724:38)
    at applySkin (https://server//include/layout.js:556:3)
    at Object.success (https://server//include/layout.js:1709:3)
    at fire (https://server//include/js/jquery.js:3305:31)
    at Object.fireWith [as resolveWith] (https://server//include/js/jquery.js:3435:7)
    at done (https://server//include/js/jquery.js:9242:14)
    at XMLHttpRequest.<anonymous> (https://server//include/js/jquery.js:9484:9)

layout.js:1724 https://server//graph_view.php?newtheme=1
Error
    at pushState (https://server//include/layout.js:1724:38)
    at Object.success (https://server//include/layout.js:1711:3)
    at fire (https://server//include/js/jquery.js:3305:31)
    at Object.fireWith [as resolveWith] (https://server//include/js/jquery.js:3435:7)
    at done (https://server//include/js/jquery.js:9242:14)
    at XMLHttpRequest.<anonymous> (https://server//include/js/jquery.js:9484:9)
```

**Comment by cigamit** [comment]
This is resolved in 1.1.30
